File: wasm-of-ocaml/samples/arith.ml
(* https://ocaml.org/learn/tutorials/99problems.html problems 32-34 *)
let rec gcd a b =
    if b = 0 then a else gcd b (a mod b)
	
let coprime a b = gcd a b = 1

let phi n =
    let rec count_coprime acc d =
      if d < n then
        count_coprime (if coprime n d then acc + 1 else acc) (d + 1)
      else acc
    in
    if n = 1 then 1 else count_coprime 0 1

let x = gcd 3094 3874 (* 26 *)
let y = phi 10 (* 4 *)

File: wasm-of-ocaml/benchmarks/C/arith.c
// https://ocaml.org/learn/tutorials/99problems.html problems 32-34
// imperative implementation of the same problem. Some recursion replaced with iteration
int gcd(int a, int b){
    if (b == 0) return a;
    return gcd(b, a%b);
}

int coprime(int a, int b){
    return gcd(a, b) == 1;
}

int phi(int n){
    if (n == 1) return 1;
    int acc = 0;
    for (int d = 1; d < n; d++){
        if (coprime(n, d)) acc++;
    }
    return acc;
}

int main(){
    int x = phi(20000);
    int y = 0;
    for (int i = 3000; i > 0; i--){
        y += phi(i);
    }
    return x + y; // so that above code can't be optimised away
}
